### ISSUE TITLE
Improvements to the decal system.

### ISSUE THREAD

**Issue by invertednormal** [o3de/o3de#15430]
**Is your feature request related to a problem? Please describe.**

The decal system is severely limited, to the extent where it would be difficult to use in a production environment.
- It's only possible to have 5 different sizes of decals. While it's possible to expand the number available, the way to go about doing this involves updating a shader to increase the number of cases in a switch statement - see https://github.com/o3de/o3de/blob/5d695ee9a419df9357325b816dc8c7ccb7e84b14/Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Decals.azsli#L85
- Normal maps and base color maps must be the same size. A single texture id / index pair is used to look up both the base color and normal, so they have to exist in the same spot on the same size permutation.
- All decals must include a normal map or normal maps break for all decals. This is related to the previous issue - since a single id/index pair is used for all textures, there can't be "holes" in the data. 

**Describe the solution you'd like**
- Preferably there would be no hard limits on the decals system beyond the memory and performance of available hardware.
- Textures attached to the decal material should be allowed to be arbitrary sizes
- The decal material should be allowed to include or not include various textures - this will become more important if support for other material properties is added.

**Some ideas on how to solve:**

For high end platforms, just using bindless resources seems like the most straightforward way to go. A structured buffer can contain the data for all decals, then a single decal ID would then point to various bindless texture IDs depending on the textures in use. A special texture id (like 0xFFFF) could be used to denote the texture isn't used, like if the decal material is missing a normal map.

For lower end platforms without bindless support this is much harder to solve. One idea would be to have an atlas (or array of atlases) to store the decals in. O3de currently supports an atlas for power-of-two textures, but it would be possible to build an atlas that supports arbitrary texture sizes. Care would need to be given to mip generation and padding between decals in the atlas to make sure they don't bleed into each other at lower mip levels. With this kind of scheme, each decal ID would point to some uv coorindate min/max in the altas for each texture used in the decal material.

**Comment by rgba16f** [comment]
As an alternate approach for the lower end platforms we could do a simple stencil multipass technique.

**Comment by moudgils** [comment]
Another possible approach to move to bindless->

- Modify the decal related viewsrg to hold both texture2darrays as well as indices (for all the decal textures) into the bindless heap . Remember that we are already paying the cost of storing the decal texture related descriptors in the static section of gpu descriptor heap.

- Update the forward pass shaders to support both the bindless method of sampling as well as the existing method of sampling and use shader options to switch between the two. 

- Modify c++ side to update the SRG with the bindless textures indices for the decal textures from RHI and set the correct shader option.

- To be  honest I think we could also apply a ifdef macro on top of shader option whereby the main pipeline just uses the bindless implementation and other pipelines has the shader option branch.

- Use SMC to generate expanded shadervariant list with the newly introduced bindless decal shader option via using system shader options support.

Basically the only downside is that we are paying the extra cost of keeping both texture2darrays and bindless indices around for the SRG but I think it may be worth it as it allows us to switch over to bindless giving us a lot of freedom with this feature
